nemo/export/quantize/quantizer.py

```diff
@@ -0,0 +1,189 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import tarfile
+from typing import List, Optional
+
+import ammo.torch.quantization as atq
+import torch.distributed as dist
+from ammo.torch.export import export_model_config
+from megatron.core import parallel_state
+from omegaconf import OmegaConf
+from omegaconf.omegaconf import DictConfig, open_dict
+from pytorch_lightning.trainer.trainer import Trainer
+
+from nemo.collections.nlp.models.language_modeling.megatron_gpt_model import MegatronGPTModel
+from nemo.collections.nlp.parts.nlp_overrides import NLPDDPStrategy, NLPSaveRestoreConnector
+from nemo.collections.nlp.parts.utils_funcs import torch_dtype_from_precision
+from nemo.utils import logging
+from nemo.utils.distributed import temporary_directory
+from nemo.utils.get_rank import is_global_rank_zero
+from nemo.utils.model_utils import load_config, save_artifacts
+
+QUANT_CFG_CHOICES = {
+    "int8": atq.INT8_DEFAULT_CFG,
+    "int8_sq": atq.INT8_SMOOTHQUANT_CFG,
+    "fp8": atq.FP8_DEFAULT_CFG,
+    "int4_awq": atq.INT4_AWQ_CFG,
+    "w4a8_awq": atq.W4A8_AWQ_BETA_CFG,
+}
+
+SUPPORTED_DTYPE = [16, "16", "bf16"]  # Default precision for non-quantized layers
+
+
+class Quantizer:
+
+    """
+    Post-training quantization of Nemo checkpoints.
+
+    PTQ converts selected model layers to low-precision format (e.g., INT4, FP8) for efficient serving.
+    The process consist of several steps:
+
+        1. Loading a Nemo model from disk using appropriate parallelism strategy
+        2. Calibrating the model to obtain appropriate algorithm-specific scaling factors
+        3. Producing .qnemo tarball with model config (JSON), quantized weights (safetensors)
```

Comment:
We use extracted .nemo files for llm, i.e. just directories, the idea of a .qnemo tarball probably doesn't make sense

Reply:
We could enable producing either directory or tarball depending on user choice via `model_save.endswith(".qnemo")`. This ".qnemo" was an initial suggestion as for what to pass to a Nemo Inference container. 

I agree that directories are more convenient to work with.

CC @oyilmaz-nvidia

Reply:
Both options are enabled now via https://github.com/NVIDIA/NeMo/pull/8444/commits/a14f7669f27ae7cc5297dd8ef7d5bb6f5aeada35